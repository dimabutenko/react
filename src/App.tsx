import React from 'react'
import { Footer, Header, Main } from './components'

const App: React.FC = () => (
  <div>
    <Header />
    <Main />
    <Footer />
  </div>
)

export default App
