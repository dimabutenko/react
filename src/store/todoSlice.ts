import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Filter } from './filter';
import type { RootState } from '../app/store';
import { ITodo, TodoState } from '../type';

export const initialState: TodoState = {
  todos: [],
  filter: Filter.All,
};

export const todoSlice = createSlice({
  name: 'counter',
  initialState,
  reducers: {
    add(state: TodoState, action: PayloadAction<string>) {
      state.todos.push({
        id: Math.random(),
        name: action.payload,
        completed: false,
      });
    },
    update(state: TodoState, action: PayloadAction<ITodo>) {
      const index = state.todos.map((x) => x.id).indexOf(action.payload.id, 0);
      if (index > -1) {
        state.todos[index].completed = action.payload.completed;
      }
    },
    remove(state: TodoState, action: PayloadAction<number>) {
      const index = state.todos.map((x) => x.id).indexOf(action.payload, 0);
      if (index > -1) {
        state.todos.splice(index, 1);
      }
    },
    updateFilter(state: TodoState, action: PayloadAction<Filter>) {
      state.filter = action.payload;
    },
    clearCompleted(state: TodoState) {
      state.todos = state.todos.filter((x) => !x.completed);
    },
  },
});

export const { add, update, remove, updateFilter, clearCompleted } = todoSlice.actions;
export const todoState = (state: RootState) => state.todoState;
export default todoSlice.reducer;
