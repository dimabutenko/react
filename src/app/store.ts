import { configureStore } from '@reduxjs/toolkit'
import todoReducer, { initialState } from '../store/todoSlice'
import { listenerMiddleware } from '../store/middleware'

const todoState = JSON.parse(localStorage.getItem('todoState') ?? 'null')

export const store = configureStore({
  preloadedState: {
    todoState: todoState === null ? initialState : todoState
  },
  reducer: {
    todoState: todoReducer
  },
  middleware: (getDefaultMiddleware) => [...getDefaultMiddleware(), listenerMiddleware.middleware]
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
