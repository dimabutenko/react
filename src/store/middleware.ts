import { createListenerMiddleware, isAnyOf } from '@reduxjs/toolkit'
import { todoSlice } from './todoSlice'
import { type RootState } from '../app/store'

export const listenerMiddleware = createListenerMiddleware()
listenerMiddleware.startListening({
  matcher: isAnyOf(
    todoSlice.actions.add,
    todoSlice.actions.remove,
    todoSlice.actions.update,
    todoSlice.actions.clearCompleted
  ),
  effect: (action, listenerApi) => {
    localStorage.setItem(
      'todoState',
      JSON.stringify((listenerApi.getState() as RootState).todoState)
    )
  }
})
