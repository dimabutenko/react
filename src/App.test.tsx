import React from 'react'
import { render } from '@testing-library/react'
import App from './App'
import { Provider } from 'react-redux'
import { store } from './app/store'

describe('App conpoonent tests', () => {
  test('Should render correctly', () => {
    const { container } = render(
      <Provider store={store}>
        <App />
      </Provider>
    )
    expect(container).toMatchSnapshot()
  })
})
