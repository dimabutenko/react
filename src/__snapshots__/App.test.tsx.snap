// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App conpoonent tests Should render correctly 1`] = `
<div>
  <div>
    <header
      class="card"
    >
      <h1>
        TODO
      </h1>
    </header>
    <main>
      <div
        class="card add-card"
      >
        <input
          class="input"
          placeholder="Create a new todo..."
          type="text"
        />
        <button
          class="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-circle-plus fa-2xl icon"
            data-icon="circle-plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM232 344V280H168c-13.3 0-24-10.7-24-24s10.7-24 24-24h64V168c0-13.3 10.7-24 24-24s24 10.7 24 24v64h64c13.3 0 24 10.7 24 24s-10.7 24-24 24H280v64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <div>
        <ul
          class="todos"
        />
        <div
          class="card stat"
        >
          <p
            class="corner"
          >
            <span
              id="items-left"
            >
              0
            </span>
             items left
          </p>
          <div
            class="filter"
          >
            <button
              class="on"
            >
              All
            </button>
            <button>
              Active
            </button>
            <button>
              Completed
            </button>
          </div>
          <div
            class="corner"
          >
            <button>
              Clear Completed
            </button>
          </div>
        </div>
      </div>
    </main>
    <footer>
      <p>
        Drag and drop to reorder list
      </p>
    </footer>
  </div>
</div>
`;
